utils/slsa_zero_two_converter.go

```diff
@@ -0,0 +1,124 @@
+package utils
+
+import (
+	"encoding/json"
+	"fmt"
+
+	pb "github.com/grafeas/grafeas/proto/v1/grafeas_go_proto"
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	"google.golang.org/protobuf/types/known/structpb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// ConvertSLSAZeroTwo function converts the raw intoto statement with SLSA 0.2 predicate to the
+// grafeas proto InTotoStatement where slsa 0.2 predicate is oneof the predicate types
+func ConvertSLSAZeroTwo(rawPayload []byte) (*pb.InTotoStatement, error) {
+	from := intoto.ProvenanceStatement{}
+	if err := json.Unmarshal(rawPayload, &from); err != nil {
+		return nil, err
+	}
+
+	predicate, err := convertPredicate(from)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.InTotoStatement{
+		Type:          from.Type,
+		Subject:       convertSubject(from.Subject),
+		PredicateType: from.PredicateType,
+		Predicate:     predicate,
+	}, nil
+
+}
+
+func convertPredicate(input intoto.ProvenanceStatement) (*pb.InTotoStatement_SlsaProvenanceZeroTwo, error) {
+	// convert the fields of interface{} type to *structpb.Struct
+	parameter, err := convertEmptyInterfaceToProtoStruct(input.Predicate.Invocation.Parameters)
+	if err != nil {
+		return nil, err
+	}
+
+	buildConfig, err := convertEmptyInterfaceToProtoStruct(input.Predicate.BuildConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	environment, err := convertEmptyInterfaceToProtoStruct(input.Predicate.Invocation.Environment)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.InTotoStatement_SlsaProvenanceZeroTwo{
+		SlsaProvenanceZeroTwo: &pb.SlsaProvenanceZeroTwo{
```

Comment:
Yes there are more differences so we need to manually convert them.
Specifically, [some json field names of predicate in intoto-golang](https://github.com/in-toto/in-toto-golang/blob/07976bf395d5fbc5030b4a576579ee70cd64cf88/in_toto/slsa_provenance/v0.2/provenance.go#L11-L18) are camelCase, whereas [the corresponding fields in grafeas proto](https://github.com/grafeas/grafeas/blob/bc2e0363cdefcb36f4f5fa20bf45e7ff9f588030/proto/v1/grafeas_go_proto/slsa_provenance_zero_two.pb.go#L39-L50) are underscore_separated_names since grafeas proto is following [proto style guide](https://developers.google.com/protocol-buffers/docs/style#message_and_field_names) to use unsersore_separated_names.

Other than that, some fields in grafeas are of type `*_struct.Struct`, and the correspondings are of type `interface{}` i.e. `BuildConfig` in predicate.

Reply:
IIRC, proto3 will accept both camel and underscore style names during unmarshalling, right? so these are only issues if we're going from grafeas -> intoto?

Reply:
Just tried to do direct unmarshal on the predicate part. It skips all the underscore fields i.e.  `build_config`, `build_type`, and some metadata fields named with underscore style in proto.